presto-tests/src/test/java/com/facebook/presto/tests/TestHistoryBasedRetry.java

```diff
@@ -0,0 +1,222 @@
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import static com.facebook.presto.SystemSessionProperties.JOIN_DISTRIBUTION_TYPE;
+import static com.facebook.presto.SystemSessionProperties.JOIN_REORDERING_STRATEGY;
+import static com.facebook.presto.SystemSessionProperties.QUERY_RETRY_LIMIT;
+import static com.facebook.presto.SystemSessionProperties.RESTRICT_HISTORY_BASED_OPTIMIZATION_TO_COMPLEX_QUERY;
+import static com.facebook.presto.SystemSessionProperties.RETRY_QUERY_WITH_HISTORY_BASED_OPTIMIZATION;
+import static com.facebook.presto.SystemSessionProperties.TRACK_HISTORY_BASED_PLAN_STATISTICS;
+import static com.facebook.presto.SystemSessionProperties.TRACK_HISTORY_STATS_FROM_FAILED_QUERIES;
+import static com.facebook.presto.SystemSessionProperties.USE_HISTORY_BASED_PLAN_STATISTICS;
+import static com.facebook.presto.SystemSessionProperties.USE_PERFECTLY_CONSISTENT_HISTORIES;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static org.testng.Assert.assertEquals;
+
+public class TestHistoryBasedRetry
+{
+    private ListeningExecutorService executor;
+
+    @BeforeClass(alwaysRun = true)
+    public void setUp()
+    {
+        executor = MoreExecutors.listeningDecorator(newCachedThreadPool());
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void shutdown()
+    {
+        executor.shutdownNow();
+    }
+
+    @Test
+    public void testQueryRetryWithHBONaturalFail()
+            throws Exception
+    {
+        int totalQueries = 1;
+        int retryLimit = 3;
+        int retryNum = 0;
+
+        String sql = "select o.orderkey, l.partkey, l.mapcol[o.orderkey] from (select orderkey, partkey, mapcol " +
+                "from (select *, map(array[1], array[2]) mapcol from lineitem)) l " +
+                "join orders o on l.partkey=o.custkey where length(comment)>10";
+
+        Session session = Session.builder(createSession())
+                .setSystemProperty(QUERY_RETRY_LIMIT, String.valueOf(retryLimit))
+                .build();
+
+        try (QueryRunner queryRunner = createQueryRunner(session)) {
+            List<ListenableFuture<?>> queryFutures = new ArrayList<>();
+            for (int i = 0; i < totalQueries; i++) {
+                ListenableFuture<?> future = executor.submit(() -> queryRunner.execute(session, sql));
+                queryFutures.add(future);
+            }
+            waitForQueryToFinish(queryFutures);
+
+            DistributedQueryRunner x = (DistributedQueryRunner) queryRunner;
+            List<BasicQueryInfo> queryInfos = x.getCoordinator().getQueryManager().getQueries();
+
+            for (BasicQueryInfo info : queryInfos) {
+                System.out.println(info.getErrorCode());
```

Comment:
delete this line; never call System.out.println in production or tests; only when debugging